README.md

[![Arduino CI](https://github.com/RobTillaart/AM2315/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/AM2315/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/AM2315/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/AM2315/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/AM2315/actions/workflows/jsoncheck.yml)
[![GitHub issues](https://img.shields.io/github/issues/RobTillaart/AM2315.svg)](https://github.com/RobTillaart/AM2315/issues)

[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/AM2315/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/AM2315.svg?maxAge=3600)](https://github.com/RobTillaart/AM2315/releases)
[![PlatformIO Registry](https://badges.registry.platformio.org/packages/robtillaart/library/AM2315.svg)](https://registry.platformio.org/libraries/robtillaart/AM2315)


# AM2315

Arduino library for an AM2315 I2C temperature and humidity sensor.


## Description

The AM2315 is a sensor similar to the DHT12 with an I2C interface.

Although in theory this could enable multiple sensors on one bus
the AM2315 has a fixed address **0x5C** so one need to implement a 
multiplexing strategy to have multiple sensors in practice. 
See multiplexing below.

The AM2315 can also be read with the https://github.com/RobTillaart/AM232X library as it uses the same protocol. 
The AM232X library allows to read some internal registers.

#### AM2315C

The AM2315C ( note the C ) is a different sensor with a different protocol.
Check - https://github.com/RobTillaart/AM2315C
The C-version has a fixed address of **0x38** so easy to detect.


#### Typical parameters

|             |  range     | accuracy | repeatability |
|:------------|:----------:|:--------:|:-------------:|
| Temperature | -40 - 125  | 0.5°C    |  ±0.2         |
| Humidity    | 0.0 - 99.9 | ±2%      |  ±0.1         |
| Sample time | 2 seconds  |          |               |


#### Hardware connection

```
//  AM232X PIN layout             AM2315 COLOR
//  ============================================
//   bottom view  DESCRIPTION     COLOR
//       +---+
//       |o  |       VDD          RED
//       |o  |       SDA          YELLOW
//       |o  |       GND          BLACK
//       |o  |       SCL          GREY
//       +---+
//
// do not forget pull up resistors between SDA, SCL and VDD.
```


#### I2C clock speed

The datasheet states the AM2315 should be used on 100 KHz I2C only. 
When overclocking I got good readings up to 190 KHz in a test with 
- Arduino UNO
- very short wires ( < 1 meter)
- not using pull ups.
- version 0.1.1 of this library


| I2C clock | timing us | Notes                 |
|:---------:|:---------:|:----------------------|
|   50 KHz  |    4570   | under-clocking works (e.g. long wires)
|  100 KHz  |    3276   | specs default, robust
|  150 KHz  |    2836   |
|  160 KHz  |    2792   |
|  170 KHz  |    2750   | 0.5 ms off, interesting for performance.
|  180 KHz  |    2700   | near critical. DO NOT USE.
|  190 KHz  |    2672   | near critical. DO NOT USE.
|  200 KHz  |   crash   | sensor needs a power cycle reboot. DO NOT USE.


If robustness is mandatory stick to the default of 100 KHz.
If performance is mandatory do not go beyond 170 KHz.


## Interface

```cpp
#include "AM2315.h"
```


#### Constructor

- **AM2315(TwoWire \*wire = &Wire)** constructor, default using Wire (I2C bus), optionally set to Wire0 .. WireN.
- **bool begin(uint8_t dataPin, uint8_t clockPin)** begin for ESP32 et al, to set I2C bus pins.
Returns true if device address 0x5C is connected.
- **bool begin()** initializer for non ESP32 e.g. AVR.
Returns true if device address 0x5C is connected.
- **bool isConnected(uint16_t timeout = 3000)** returns true if the device address 0x5C is found on I2C bus.
As the device can be in sleep modus it will retry for the defined timeout (in micros) with a minimum of 1 try. 
minimum = 800 us and maximum = 3000 us according to datasheet.


#### Core

- **int8_t read()** read the sensor and store the values internally.
It returns the status of the read which should be **AM2315_OK** == 0.
- **float getHumidity()** returns last read humidity + optional offset, 
or **AM2315_INVALID_VALUE** == -999 in case of error. 
This error can be suppressed, see below.
- **float getTemperature()** returns last read temperature + optional offset,
or **AM2315_INVALID_VALUE** == -999 in case of error. 
This error can be suppressed, see below.
- **uint32_t lastRead()** returns the timestamp in milliseconds since startup of the last successful read.


#### Offset

- **void setHumOffset(float offset = 0)** set an offset for humidity to calibrate (1st order) the sensor.
Default offset = 0, so no parameter will reset the offset.
- **float getHumOffset()** return current humidity offset, default 0.
- **void setTempOffset(float offset = 0)** set an offset for temperature to calibrate (1st order) the sensor.
Default offset = 0, so no parameter will reset the offset.
- **float getTempOffset()** return current temperature offset, default 0.


#### Control

Functions to adjust the communication with the sensor.

- **void setWaitForReading(bool b )** flag to enforce a blocking wait (up to 2 seconds) when **read()** is called.
- **bool getWaitForReading()** returns the above setting.
- **bool wakeUp()** function that will try for 3 milliseconds to wake up the sensor.
This can be done before an actual read to minimize the **read()** call.
- **void setSuppressError(bool b)** suppress error values of **AM2315_INVALID_VALUE** == -999 
=> you need to check the return value of read() instead.  
This can be used to keep spikes out of your graphs / logs. 
- **bool getSuppressError()**  returns the above setting.


#### Error codes

| name                              | value | notes       |
|:----------------------------------|------:|:------------|
| AM2315_OK                         |  0    |
| AM2315_ERROR_CHECKSUM             |  -10  | I2C problem.
| AM2315_ERROR_CONNECT              |  -11  | I2C problem.
| AM2315_MISSING_BYTES              |  -12  | I2C problem.
| AM2315_WAITING_FOR_READ           |  -50  | called **read()** too fast, within 2 seconds.
| AM2315_HUMIDITY_OUT_OF_RANGE      |  -100 | not used by default.
| AM2315_TEMPERATURE_OUT_OF_RANGE   |  -101 | not used by default.
| AM2315_INVALID_VALUE              |  -999 | can be suppressed. 

Note: The OUT_OT_RANGE errors are tested on the "raw" readings, excluding the offset.
So if the "raw" humidity is 99% and you use an offset of 3% you will get 102%
without the error. 


## Operation

See examples

In **setup()** you have to call the **begin()** to initialize 
the Wire library and do an initial **read()** to fill the variables temperature and humidity. 
To access these values one must use **getTemperature()** and **getHumidity()**. 
Multiple calls will give the same values until **read()** is called again.

Note that the AM2315 can go into sleep mode after 3 seconds after last read, 
so one might need to call **wakeUp()** before the **read()**.


## Multiplexing 

Multiplexing the **AM2315** can be done in several ways.
This is not a complete list or tutorial but should get you started.

1. Control the power line by means of an extra pin (+ transistor). 
Only switch on the sensor you want to use. Drawback might be time 
the sensor takes to boot and to be ready for the first measurement.
2. Use an AND gate between the I2C SCL (clock) line and the I2C SCL 
pin of the sensors. This way one can enable / disable communication 
per sensor. This will still need an IO pin per sensor but does not 
have the "boot time" constraint mentioned above.
You may use a **PCF8574** to control the AND gates.
https://github.com/RobTillaart/PCF8574
3. Use a **TCA9548A** I2C Multiplexer, or similar. https://github.com/RobTillaart/TCA9548

Which method fit your application depends on your requirements and constraints.


## Future

#### Must

- update documentation
- test more (other platforms)

#### Should

- keep in sync with AM232X class
  - merge in a far future.
- update unit test
- add examples

#### Could

- move code from .h to .cpp

#### Wont

- add calls for meta information (no description yet)
  - 0x07 status register
  - 0x08-0x0B user register HIGH LOW HIGH2 LOW2
  (use AM232x library to access those)


## Support

If you appreciate my libraries, you can support the development and maintenance.
Improve the quality of the libraries by providing issues and Pull Requests, or
donate through PayPal or GitHub sponsors.

Thank you,



CHANGELOG.md
# Change Log AM2315

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.2.0] - 2023-10-13
- simplified **begin()** no more Wire pin setting.
- add **AM2315_Wire1.ino** example
- update readme.md
- minor edits

----

## [0.1.7] - 2023-01-21
- update GitHub actions
- update license 2023
- update keywords.txt
- minor edits comments
- add offset example 
- minor edit README.md

## [0.1.6] - 2022-10-27
- Add RP2040 support to build-CI.
- Add CHANGELOG.md

## [0.1.5] - 2020-03-25
- fix #11 Warning requestFrom for ESP32

## [0.1.4] - 2022-06-18
- sync with AM232X (readme)

## [0.1.3] - 2022-06-17
-  sync with AM232X

## [0.1.2] - 2022-01-13
-  fix wake-up in read() for ESP32

## [0.1.1] - 2022-01-11
-  fix handshake

## [0.1.0] - 2022-01-05
-  initial version



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2022-01-05
// PURPOSE: unit tests for the AM2315 temperature and humidity sensor
//          https://github.com/RobTillaart/AM2315
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// https://github.com/Arduino-CI/arduino_ci/blob/master/cpp/unittest/Assertion.h#L33-L42
// ----------------------------
// assertEqual(expected, actual)
// assertNotEqual(expected, actual)
// assertLess(expected, actual)
// assertMore(expected, actual)
// assertLessOrEqual(expected, actual)
// assertMoreOrEqual(expected, actual)
// assertTrue(actual)
// assertFalse(actual)
// assertNull(actual)
// assertNotNull(actual)


#include <ArduinoUnitTests.h>


#include "Wire.h"
#include "AM2315.h"


unittest_setup()
{
  fprintf(stderr, "AM2315_LIB_VERSION: %s \n", (char *) AM2315_LIB_VERSION);
}


unittest_teardown()
{
}


unittest(test_constants)
{
  assertEqual(  0, AM2315_OK);
  assertEqual(-10, AM2315_ERROR_CHECKSUM);
  assertEqual(-11, AM2315_ERROR_CONNECT);
  assertEqual(-12, AM2315_MISSING_BYTES);
  assertEqual(-50, AM2315_WAITING_FOR_READ);

  assertEqual(-100, AM2315_HUMIDITY_OUT_OF_RANGE);
  assertEqual(-101, AM2315_TEMPERATURE_OUT_OF_RANGE);
  assertEqual(-999, AM2315_INVALID_VALUE);
}


unittest(test_constructor)
{
  AM2315 sensor(&Wire);

  assertEqualFloat(0, sensor.getTemperature(), 0.001);
  assertEqualFloat(0, sensor.getHumidity(), 0.001);
  assertEqualFloat(0, sensor.getTempOffset(), 0.001);
  assertEqualFloat(0, sensor.getHumOffset(), 0.001);

  Wire.begin();
  sensor.begin();

  assertEqual(AM2315_WAITING_FOR_READ, sensor.read());

  assertEqualFloat(0, sensor.getTemperature(), 0.001);
  assertEqualFloat(0, sensor.getHumidity(), 0.001);
  assertEqualFloat(0, sensor.getTempOffset(), 0.001);
  assertEqualFloat(0, sensor.getHumOffset(), 0.001);
}


unittest(test_offset)
{
  AM2315 sensor(&Wire);

  assertEqualFloat(0, sensor.getTempOffset(), 0.001);
  assertEqualFloat(0, sensor.getHumOffset(), 0.001);

  for (float offset = -1.5; offset < 1.5; offset += 0.5)
  {
    sensor.setHumOffset(offset);
    sensor.setTempOffset(offset);

    assertEqualFloat(offset, sensor.getTempOffset(), 0.001);
    assertEqualFloat(offset, sensor.getHumOffset(), 0.001);

    assertEqualFloat(offset, sensor.getTemperature(), 0.001);
    assertEqualFloat(offset, sensor.getHumidity(), 0.001);
  }
}


unittest_main()


//  -- END OF FILE --



AM2315.cpp
//
//    FILE: AM2315.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.7
// PURPOSE: AM2315 Temperature and Humidity sensor library for Arduino
//     URL: https://github.com/RobTillaart/AM2315
//
//  HISTORY: see changelog.md


#include "AM2315.h"


const uint8_t AM2315_ADDRESS = 0x5C;

//  READ_DELAY for blocking read
const uint16_t AM2315_READ_DELAY = 2000;


/////////////////////////////////////////////////////
//
//  PUBLIC
//
AM2315::AM2315(TwoWire *wire)
{
  _wire          = wire;
  //  reset() or begin() ?
  _humidity      = 0.0;
  _temperature   = 0.0;
  _humOffset     = 0.0;
  _tempOffset    = 0.0;
  _lastRead      = 0;
  _waitForRead   = false;
  _suppressError = false;
}


#if defined(ESP8266) || defined(ESP32)
bool AM2315::begin(const uint8_t dataPin, const uint8_t clockPin)
{
  if ((dataPin < 255) && (clockPin < 255))
  {
    _wire->begin(dataPin, clockPin);
  } else {
    _wire->begin();
  }
  if (! isConnected()) return false;
  this->read();
  return true;
}
#endif


bool AM2315::begin()
{
  _wire->begin();
  if (! isConnected()) return false;
  this->read();
  return true;
}


bool AM2315::isConnected(uint16_t timeout)
{
  uint32_t start = micros();
  while (micros() - start < timeout)
  {
    _wire->beginTransmission(AM2315_ADDRESS);
    if ( _wire->endTransmission() == 0) return true;
    yield();
    delayMicroseconds(100);
  }
  return false;
}


//  return values:
//    AM2315_OK
//    AM2315_ERROR_CONNECT
//    AM2315_MISSING_BYTES
//    AM2315_ERROR_CHECKSUM;
//    AM2315_HUMIDITY_OUT_OF_RANGE
//    AM2315_TEMPERATURE_OUT_OF_RANGE
int AM2315::read()
{
  while (millis() - _lastRead < AM2315_READ_DELAY)
  {
    if (!_waitForRead) return AM2315_WAITING_FOR_READ;
    yield();
  }
  int rv = _read();
  _lastRead = millis();
  return rv;
}


float AM2315::getHumidity()
{
  float _hum = _humidity;
  if (_humOffset != 0.0) _hum += _humOffset;
  return _hum;
}


float AM2315::getTemperature()
{
  float _tem = _temperature;
  if (_tempOffset != 0.0) _tem += _tempOffset;
  return _tem;
}


///////////////////////////////////////////////////////////
//
//  PRIVATE
//

//  return values:
//    AM2315_OK
//    AM2315_ERROR_CONNECT
//    AM2315_MISSING_BYTES
//    AM2315_ERROR_CHECKSUM;
//    AM2315_HUMIDITY_OUT_OF_RANGE
//    AM2315_TEMPERATURE_OUT_OF_RANGE
int AM2315::_read()
{
  // READ VALUES
  int rv = _readSensor();

  if (rv != AM2315_OK)
  {
    if (_suppressError == false)
    {
      _humidity    = AM2315_INVALID_VALUE;
      _temperature = AM2315_INVALID_VALUE;
    }
    return rv;  //  propagate error value
  }

  //  EXTRACT HUMIDITY AND TEMPERATURE
  _humidity = (_bits[2] * 256 + _bits[3]) * 0.1;
  int16_t t = ((_bits[4] & 0x7F) * 256 + _bits[5]);
  if (t == 0)
  {
    _temperature = 0.0;     // prevent -0.0;
  }
  else
  {
    _temperature = t * 0.1;
    if ((_bits[4] & 0x80) == 0x80 )
    {
      _temperature = -_temperature;
    }
  }

#ifdef AM2315_VALUE_OUT_OF_RANGE
  //  TEST OUT OF RANGE
  if (_humidity > 100)
  {
    return AM2315_HUMIDITY_OUT_OF_RANGE;
  }
  if ((_temperature < -40) || (_temperature > 125))
  {
    return AM2315_TEMPERATURE_OUT_OF_RANGE;
  }
#endif

  return AM2315_OK;
}


//  return values:
//    AM2315_OK
//    AM2315_ERROR_CONNECT
//    AM2315_MISSING_BYTES
//    AM2315_ERROR_CHECKSUM;
int AM2315::_readSensor()
{
  //  HANDLE PENDING IRQ etc.
  yield();

  // WAKE UP the sensor
  if (! isConnected() ) return AM2315_ERROR_CONNECT;

  //  SEND COMMAND
  _wire->beginTransmission(AM2315_ADDRESS);
  _wire->write(0X03);
  _wire->write(0);
  _wire->write(4);
  int rv = _wire->endTransmission();
  if (rv < 0) return rv;

  //  REQUEST DATA
  const int length = 8;
  int bytes = _wire->requestFrom((int)AM2315_ADDRESS, length);
  if (bytes == 0)     return AM2315_ERROR_CONNECT;
  if (bytes < length) return AM2315_MISSING_BYTES;

  //  READ DATA
  for (int i = 0; i < bytes; i++)
  {
    _bits[i] = _wire->read();
  }

  //  TEST CHECKSUM
  uint16_t crc = _bits[bytes - 1] * 256 + _bits[bytes - 2];
  if (_crc16(_bits, bytes - 2) != crc)
  {
    return AM2315_ERROR_CHECKSUM;
  }
  return AM2315_OK;
}


uint16_t AM2315::_crc16(uint8_t *ptr, uint8_t len)
{
  uint16_t crc = 0xFFFF;
  while(len--)
  {
    crc ^= *ptr++;
    for (uint8_t i = 0; i < 8; i++)
    {
      if (crc & 0x01)
      {
        crc >>= 1;
        crc ^= 0xA001;
      }
      else
      {
        crc >>= 1;
      }
    }
  }
  return crc;
}


//  -- END OF FILE --



AM2315.h
#pragma once
//
//    FILE: AM2315.h
//  AUTHOR: Rob Tillaart
// PURPOSE: AM2315 Temperature and Humidity sensor library for Arduino
// VERSION: 0.1.2
//     URL: https://github.com/RobTillaart/AM2315
//
//  AM232X PIN layout             AM2315 COLOR
//  ============================================
//   bottom view  DESCRIPTION     COLOR
//       +---+
//       |o  |       VDD          RED
//       |o  |       SDA          YELLOW
//       |o  |       GND          BLACK
//       |o  |       SCL          GREY
//       +---+
//
// do not forget pull up resistors between SDA, SCL and VDD.


#include "Arduino.h"
#include "Wire.h"


#define AM2315_LIB_VERSION                    (F("0.1.2"))


#define AM2315_OK                             0
#define AM2315_ERROR_CHECKSUM                 -10
#define AM2315_ERROR_CONNECT                  -11
#define AM2315_MISSING_BYTES                  -12
#define AM2315_WAITING_FOR_READ               -50


//  optionally detect out of range values.
//  occurs seldom so not enabled by default.
//  #define AM2315_VALUE_OUT_OF_RANGE
#define AM2315_HUMIDITY_OUT_OF_RANGE          -100
#define AM2315_TEMPERATURE_OUT_OF_RANGE       -101

//  allows to overrule AM2315_INVALID_VALUE e.g. to prevent spike in graphs.
#ifndef AM2315_INVALID_VALUE
#define AM2315_INVALID_VALUE                  -999
#endif


class AM2315
{
public:

  AM2315(TwoWire *wire = &Wire);

#if defined(ESP8266) || defined(ESP32)
  bool     begin(const uint8_t dataPin, const uint8_t clockPin);
#endif
  bool     begin();
  // datasheet- wake up is min 800 us max 3000 us
  bool     isConnected(uint16_t timeout = 3000);

  int      read();

  // lastRead is in MilliSeconds since start sketch
  uint32_t lastRead()                    { return _lastRead; };

  // preferred interface
  float    getHumidity();
  float    getTemperature();

  // adding offsets works well in normal range
  // might introduce under- or overflow at the ends of the sensor range
  void     setHumOffset(float offset)    { _humOffset = offset; };
  void     setTempOffset(float offset)   { _tempOffset = offset; };
  float    getHumOffset()                { return _humOffset; };
  float    getTempOffset()               { return _tempOffset; };

  bool     getWaitForReading()           { return _waitForRead; };
  void     setWaitForReading(bool b )    { _waitForRead = b; };


  // suppress error values of -999 => check return value of read() instead
  bool     getSuppressError()            { return _suppressError; };
  void     setSuppressError(bool b)      { _suppressError = b; };

  bool     wakeUp() { return isConnected(); };

private:
  float    _humOffset     = 0.0;
  float    _tempOffset    = 0.0;
  float    _humidity      = 0.0;
  float    _temperature   = 0.0;
  uint32_t _lastRead      = 0;
  bool     _waitForRead   = false;
  bool     _suppressError = false;

  uint8_t  _bits[8];    // buffer to hold raw data
  int      _read();
  int      _readSensor();
  uint16_t _crc16(uint8_t *ptr, uint8_t len);

  TwoWire* _wire;
};


// -- END OF FILE --

